Automattic/sensei
# Add Sensei Home plugin update notices to REST API

Fixes #5738 

### Changes proposed in this Pull Request

* Adds a provider for notices to inject notices into Sensei Home's REST API endpoint.
* Adds a handler for injecting special Sensei Home notices either into `Sensei_Admin_Notices` (when available) or sneakily through the shared filter in the provider when `Sensei_Admin_Notices` isn't available (Sensei Blocks).
* Begrudgingly makes `Sensei_Admin_Notices` a singleton. It has a mix of classes that depend on it and would require several layers of DI. 
* We planned on having this be called `notifications` but switched to `notices` to match the rest of Sensei.
* Adds the following notices:
  * When an update is available for a free plugin or licensed and activated paid plugin.
  * When an update is available for a paid plugin that isn't active an we're not sure the license status.
  * When an update is available for a paid plugin but it doesn't have an active license.

Note: This doesn't add any additional license notices yet (when a license is about to expire, for example). It was getting too big so I thought I'd wait.

### Testing instructions
* Try the following with various situations: Certificates downgraded, Sensei Pro deactivated, etc.
  * You can also try this snippet to see what it is like with licensed Sensei Pro (future PR): 
  * `add_filter( 'sensei_home_is_plugin_licensed_sensei-pro', '__return_true' );`
* Run: `curl -s -X GET --user jake:jakejake 'http://sensei.local/wp-json/sensei-internal/v1/home' | jq .notices`
* Confirm output makes sense.

### New/Updated Hooks

* `sensei_show_admin_notices_{$screen_id}`: Allows hiding admin notices on a particular screen. Used to hide on Sensei Home.
* `sensei_home_is_plugin_licensed_{$plugin_slug}`:  Allows Sensei Pro and Sensei Blocks to tell Sensei Home when their plugins have an active license.
